datalad/acmd.py

```diff
@@ -0,0 +1,131 @@
+# emacs: -*- mode: python; py-indent-offset: 4; tab-width: 4; indent-tabs-mode: nil -*-
+# ex: set sts=4 ts=4 sw=4 noet:
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+#
+#   See COPYING file distributed along with the datalad package for the
+#   copyright and license terms.
+#
+# ## ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ### ##
+"""Interface for running commands with `asyncio.create_subprocess_{shell,exec}`.
+
+This is intended for internal use by `cmd.Runner`.
+"""
+
+import asyncio
+from functools import partial
+import locale
+import logging
+import sys
+
+lgr = logging.getLogger('datalad.acmd')
+
+
+async def _noop():
+    return None
+
+
+async def _read_stream(stream, callback):
+    while True:
+        line = await stream.readline()
+        if line:
+            callback(line)
+        else:
```

Comment:
As far as I can see an empty line would stop the stream reading. I guess that's unintended and the condition should be sth like `if line is not None`?

Reply:
> As far as I can see an empty line would stop the stream reading.

I think the usual "empty string marks EOF" applies.

<details>
<summary>quick test</summary>

```python
from datalad.cmd import Runner
cmd = ("printf 'a' && "
       "printf '' && "       # empty
       "printf '\n' && "     # blank line
       "printf 'err' >&2 &&"
       "sleep 1 && "
       "printf 'c\n'")
ret = Runner().run(cmd, shell=True, log_online=True, log_stderr=True)
print("====================================================")
print(f"return: {ret}")
```

```
====================================================
return: ('a\nc\n', 'err')
```

</details>

Reply:
FWIW, I was interested in "online callback" behavior.  Apparently we don't even bother about empty lines currently.  And there is a tiny difference in behavior with this branch:
<details>
<summary>my tuned version of the script (sorry for mixed tabulation ;))</summary> 

```python
from time import time

from datalad.cmd import Runner                                                                                                  
cmd = (
    "printf 'a' && "                                                                                                         
    "sleep 0.1 && "
	"printf '' && "       # empty                                                                                            
    "sleep 0.1 && "
	"printf '\n' && "     # blank line                                                                                       
    "sleep 0.1 && "
	"printf '\r' && "     # CR
    "sleep 0.1 && "
	"printf 'err' >&2 &&"                                                                                                    
	"sleep 1 && "                                                                                                            
	"printf 'c\n'")                                                                                                          

t0 = time()
def online(line):
   dt = time() - t0
   print("%6.3f Got: %r" % (dt, line))

ret = Runner().run(cmd, shell=True, log_stdout=online, log_online=True, log_stderr=True)                                                           
print("====================================================")                                                                   
print(f"return: {ret}") 
```
</details>

on 0.11.x and master:
```shell
$> datalad/venvs/dev3/bin/python trash/test-runner-newline.py 
 0.204 Got: 'a\n'
 1.409 Got: '\rc'
====================================================
return: ('', 'err')
```
on this branch
```shell
$> datalad-master/venvs/dev3/bin/python trash/test-runner-newline.py 
 0.221 Got: 'a\n'
 1.425 Got: '\rc\n'
====================================================
return: ('', 'err')
```
so there is a trailing `\n` in the line with `\r`.
Also interestingly currently we do not process lines with `\r` on their own so I guess we would be missing any "scrolling" output from other tools.

Reply:
[ AFAICS this comment isn't really related to the discussion here.  To be clear: My "I think" above was a bit understated.  Using None is wrong and would cause th loop to hang indefinitely waiting for a value that would not come. ]

> so there is a trailing `\n` in the line with `\r`.

Yeah, I noticed too that our current Runner incorrectly eats some newlines, though I havne't looked into where that happens.

> Also interestingly currently we do not process lines with `\r` on their own so I guess we would be missing any "scrolling" output from other tools.

I'm not sure there is anything to do here.  I think it is fair that the callables get what is returned by readline().  What they do with it is up to them.

Reply:
> Using None is wrong and would cause the loop to hang indefinitely waiting for a value that would not come.

Yeah, that's actually quite obvious. Clearly I was too quick to comment. Furthermore even my recollection how a string evaluates to an bool seems to be outdated (at least - may be it has always been wrong). So - ignore me.